File: bridge/src/timeline-store.ts
/**
 * Bridge-side timeline store — minimal server-side event buffer.
 * Stores recent timeline entries for relay to Android/plugin clients.
 * No scroll/grouping logic (that's client-side).
 */

import type { TimelineEntry } from './types.js';
import { deduplicateEntry, normalizeTimelineEntryForStorage } from '@agentdeck/shared';
import { readFileSync } from 'fs';

type EntryListener = (entry: TimelineEntry, upsert?: boolean) => void;
/** Attribute an entry with session-scoped metadata (sessionId, projectName,
 *  taskId, runId, ...) before it lands in the buffer. Run once at storage
 *  time so history replay carries the same attribution as the live broadcast.
 *  Must be idempotent: caller already-set fields take precedence. */
type EntryAttributor = (entry: TimelineEntry) => TimelineEntry;

const MAX_ENTRIES = 200;
/** Separate retention cap for task hierarchy rows (task_start/task_end/
 *  task_milestone). Task rows are exempt from the generic FIFO shift —
 *  a long task's `task_start` must not scroll away while its turns stream
 *  in, or the eventual `task_end` renders as an unpaired orphan. ~30 pairs. */
const MAX_TASK_ENTRIES = 60;

function isTaskRow(e: Pick<TimelineEntry, 'type'>): boolean {
  return e.type === 'task_start' || e.type === 'task_end' || e.type === 'task_milestone';
}

/** Chat/tool entry types that, in projection mode, come from the SessionSample
 *  projection instead of the adapters' direct emitters. Locally-emitted entries
 *  of these types are suppressed when projection mode is on so the timeline has
 *  exactly one source. Task hierarchy + error/scheduled rows are never affected. */
const PROJECTED_TYPES: ReadonlySet<string> = new Set([
  'chat_start', 'chat_response', 'chat_end', 'tool_request', 'tool_resolved', 'tool_exec',
]);

export class BridgeTimelineStore {
  private entries: TimelineEntry[] = [];
  private listeners: EntryListener[] = [];
  private attributor: EntryAttributor | null = null;
  /** Phase 6 cutover (default OFF). When true, locally-emitted chat/tool rows
   *  are dropped — the SessionSample projection (added via `bypassSuppression`)
   *  becomes the single source. Relayed + projected entries bypass this. */
  private suppressLocalChatTool = false;

  setSuppressLocalChatTool(v: boolean): void {
    this.suppressLocalChatTool = v;
  }

  /** Install (or replace) the attributor invoked on every addEntry / upsertEntry
   *  before dedup. Wired by `BridgeCore.wireTimeline` so the store and the
   *  live broadcast see identical, fully-attributed entries — critical for
   *  history replay (`timeline_history`). */
  setAttributor(fn: EntryAttributor | null): void {
    this.attributor = fn;
  }

  addEntry(entry: TimelineEntry, opts?: { bypassSuppression?: boolean }): void {
    // Phase 6: in projection mode, drop locally-emitted chat/tool rows — they
    // now come from the SessionSample projection (which bypasses) or are
    // relayed from another bridge (which bypasses). No-op when mode is off.
    if (this.suppressLocalChatTool && !opts?.bypassSuppression && PROJECTED_TYPES.has(entry.type)) {
      return;
    }
    const enriched = this.attributor ? this.attributor(entry) : entry;
    const normalized = normalizeTimelineEntryForStorage(enriched);
    if (!normalized) return;

    if (this.tryMergeTaskEndByTaskId(normalized)) {
      return;
    }

    const result = deduplicateEntry(normalized, this.entries);

    if (result.action === 'skip') return;

    if (result.action === 'merge') {
      // Merge path is an *update* — the existing entry was attributed at its
      // own creation time and that attribution is the truth. Do NOT pull
      // session/task/run/project from `enriched` (the attributor on this call
      // reflects the *current* active task, which may have rotated since the
      // original entry was stored). Use the raw caller `entry` so explicit
      // caller-set fields still win, and otherwise keep `existing.*`.
      const existing = this.entries[result.index];
      existing.repeatCount = (existing.repeatCount || 1) + 1;
      existing.ts = normalized.ts;
      existing.raw = normalized.raw;
      existing.detail = normalized.detail ?? existing.detail;
      existing.agentType = entry.agentType ?? existing.agentType;
      existing.projectName = entry.projectName ?? existing.projectName;
      existing.sessionId = entry.sessionId ?? existing.sessionId;
      existing.runId = entry.runId ?? existing.runId;
      existing.taskId = entry.taskId ?? existing.taskId;
      existing.startedAt = normalized.startedAt ?? existing.startedAt;
      existing.endedAt = normalized.endedAt ?? existing.endedAt;
      existing.automated = normalized.automated ?? existing.automated;
      existing.summaryKind = normalized.summaryKind ?? existing.summaryKind;
      existing.boundarySignal = normalized.boundarySignal ?? existing.boundarySignal;
      existing.taskScore = normalized.taskScore ?? existing.taskScore;
      existing.taskOutcome = normalized.taskOutcome ?? existing.taskOutcome;
      existing.taskCategory = normalized.taskCategory ?? existing.taskCategory;
      existing.taskSummary = normalized.taskSummary ?? existing.taskSummary;
      if (result.removeChatStartIndex != null) {
        this.entries.splice(result.removeChatStartIndex, 1);
      }
      for (const cb of this.listeners) cb(existing, true);
      return;
    }

    // action === 'add' — fresh entry. Push the enriched form so storage and
    // history replay carry attribution from the time of creation.
    this.entries.push(result.entry);
    if (this.entries.length > MAX_ENTRIES) {
      this.evictOne();
    }
    for (const cb of this.listeners) cb(result.entry);
  }

  /** Evict a single entry to enforce MAX_ENTRIES, protecting task rows.
   *
   *  Task rows only leave under their own cap (MAX_TASK_ENTRIES), and an
   *  in-flight task's `task_start` (no matching `task_end` yet) is never
   *  evicted — otherwise a long task's start scrolls away mid-task and the
   *  pair splits. Everything else FIFOs as before. */
  private evictOne(): void {
    const taskRowCount = this.entries.reduce((n, e) => n + (isTaskRow(e) ? 1 : 0), 0);
    if (taskRowCount > MAX_TASK_ENTRIES) {
      const closed = new Set<string>();
      for (const e of this.entries) {
        if (e.type === 'task_end' && e.taskId) closed.add(e.taskId);
      }
      const idx = this.entries.findIndex(
        (e) => isTaskRow(e) && !(e.type === 'task_start' && e.taskId && !closed.has(e.taskId)),
      );
      if (idx >= 0) {
        this.entries.splice(idx, 1);
        return;
      }
    }
    // Shed the oldest `tool_exec` before any chat/turn row. Observed codex/
    // opencode command rows carry real sessionIds + command detail, so they
    // pass the anonymous-noise storage filter and accumulate — a live daemon
    // snapshot ran 87% tool_exec (87/100). Undifferentiated FIFO would then
    // evict a turn's `chat_start` (always the oldest ts in its turn) before its
    // own tool_exec rows, leaving an orphaned chat_response/chat_end in the
    // `timeline_history` replay a reconnecting client receives — the Node-side
    // face of the "answer with no prompt" symptom. tool_exec is standalone
    // (unlike the tool_request/tool_resolved approval pair, which must not be
    // split), so shedding it first is safe and preserves the turn skeleton.
    const toolIdx = this.entries.findIndex((e) => e.type === 'tool_exec');
    if (toolIdx >= 0) {
      this.entries.splice(toolIdx, 1);
      return;
    }
    const idx = this.entries.findIndex((e) => !isTaskRow(e));
    if (idx >= 0) {
      this.entries.splice(idx, 1);
      return;
    }
    // Pathological: buffer is 100% task rows — fall back to plain FIFO so the
    // buffer stays bounded.
    this.entries.shift();
  }

  getHistory(since?: number): TimelineEntry[] {
    const entries = since
      ? this.entries.filter((e) => e.ts > since)
      : this.entries;
    return [...entries].sort((a, b) => a.ts - b.ts);
  }

  /** Load persisted timeline rows into the bounded replay buffer without
   *  broadcasting them as live events. The Node daemon normally keeps the
   *  device timeline in memory, but macOS/previous daemon runs may leave a
   *  timeline.json behind. Rehydrating it lets reconnecting Android/tablet
   *  clients receive a non-empty initial `timeline_history`.
   */
  loadPersistedFile(path: string): number {
    let parsed: unknown;
    try {
      parsed = JSON.parse(readFileSync(path, 'utf-8'));
    } catch {
      return 0;
    }
    const rows = Array.isArray(parsed)
      ? parsed
      : parsed && typeof parsed === 'object' && Array.isArray((parsed as { entries?: unknown }).entries)
        ? (parsed as { entries: unknown[] }).entries
        : [];
    const before = this.entries.length;
    this.loadPersistedEntries(rows);
    return Math.max(0, this.entries.length - before);
  }

  loadPersistedEntries(rows: unknown[]): void {
    const normalized: TimelineEntry[] = [];
    for (const row of rows) {
      if (!row || typeof row !== 'object') continue;
      const entry = row as Partial<TimelineEntry>;
      if (typeof entry.ts !== 'number' || typeof entry.type !== 'string' || typeof entry.raw !== 'string') continue;
      const clean = normalizeTimelineEntryForStorage(entry as TimelineEntry);
      if (clean) normalized.push(clean);
    }
    if (normalized.length === 0) return;
    const byKey = new Map<string, TimelineEntry>();
    for (const entry of [...this.entries, ...normalized].sort((a, b) => a.ts - b.ts)) {
      byKey.set(`${entry.ts}:${entry.type}:${entry.raw}`, entry);
    }
    const merged = Array.from(byKey.values()).sort((a, b) => a.ts - b.ts);
    // Trim with the same task-row protection as live eviction: task rows keep
    // their own (deeper) retention so start/end pairs survive a reload even
    // when chat/tool rows overflowed the generic cap.
    const taskRows = merged.filter(isTaskRow).slice(-MAX_TASK_ENTRIES);
    const rest = merged.filter((e) => !isTaskRow(e)).slice(-(MAX_ENTRIES - taskRows.length));
    this.entries = [...taskRows, ...rest].sort((a, b) => a.ts - b.ts);
  }

  /** Node mirror of the Swift daemon's orphan reaper
   *  (DaemonServer.computeOrphanTaskEnds): synthesize a `task_end` for every
   *  `task_start` whose pair was never written. The producer only guarantees
   *  the pair within a single daemon lifetime (ApmeCollector.closeTask is
   *  in-memory), so a daemon killed mid-task leaves task_start rows that
   *  clients render as in-flight — spinning the task marker forever. Called
   *  once after `loadPersistedFile` at daemon startup. Idempotent: a real
   *  task_end arriving later merges over the synthetic via
   *  `tryMergeTaskEndByTaskId`. Returns the number of synthesized rows. */
  reapOrphanTaskStarts(): number {
    const closed = new Set<string>();
    for (const e of this.entries) {
      if (e.type === 'task_end' && e.taskId) closed.add(e.taskId);
    }
    let count = 0;
    for (const start of [...this.entries]) {
      if (start.type !== 'task_start' || !start.taskId || closed.has(start.taskId)) continue;
      const startedAtMs = start.startedAt ?? start.ts;
      // ts nudged 1ms past the task_start so the synthetic end sorts right
      // after the orphaned start. endedAt stays unset — the duration is
      // genuinely unknown, and clients render "Interrupted · –" for that.
      this.addEntry({
        ts: startedAtMs + 1,
        type: 'task_end',
        raw: 'Interrupted · –',
        ...(start.agentType ? { agentType: start.agentType } : {}),
        ...(start.projectName ? { projectName: start.projectName } : {}),
        ...(start.sessionId ? { sessionId: start.sessionId } : {}),
        ...(start.runId ? { runId: start.runId } : {}),
        startedAt: startedAtMs,
        taskId: start.taskId,
        boundarySignal: 'interrupted',
      }, { bypassSuppression: true });
      closed.add(start.taskId);
      count++;
    }
    return count;
  }

  /** Recent entries attributed to one session, for the `query_session_timeline`
   *  poll — lets a device that connects mid-session fill its Detail view. */
  getHistoryForSession(sessionId: string, since?: number, limit = 16): TimelineEntry[] {
    // sessions_list ids for passively-observed sessions are prefixed
    // ("observed:claude:<uuid>") while timeline entries are keyed by the raw uuid,
    // so accept either form.
    const raw = sessionId.replace(/^observed:(?:claude|codex|opencode|antigravity):/, '');
    const matched = this.entries.filter(
      (e) => (e.sessionId === sessionId || e.sessionId === raw) && (since == null || e.ts > since),
    );
    matched.sort((a, b) => a.ts - b.ts);
    // `limit` applies to the chat/tool stream; task hierarchy rows ride along
    // in full so a long task's start/end pair never splits at the per-session
    // window (a task_start older than the last `limit` rows used to vanish,
    // leaving the Detail view an unpaired task_end).
    const taskRows = matched.filter(isTaskRow);
    const rest = matched.filter((e) => !isTaskRow(e)).slice(-limit);
    return [...taskRows, ...rest].sort((a, b) => a.ts - b.ts);
  }

  updateEntryStatus(approvalId: string, status: 'approved' | 'denied'): void {
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (this.entries[i].approvalId === approvalId) {
        this.entries[i] = { ...this.entries[i], status };
        return;
      }
    }
  }

  /** Update existing entry with same ts+type (1s tolerance), or add new.
   *
   *  Update path: the existing entry's session/task/run/project attribution
   *  was set at *its* creation time and is authoritative. We do NOT re-run
   *  the attributor on update — by the time a late upsert lands (e.g. async
   *  LLM summary on a chat_end), the active task may have rotated, and
   *  re-attributing would silently move the entry to the wrong task. The
   *  raw caller `entry` only overrides existing fields when the caller
   *  explicitly set them.
   *
   *  Insert path (no match): falls through to `addEntry`, which runs the
   *  attributor as usual to capture creation-time attribution. */
  upsertEntry(entry: TimelineEntry, opts?: { bypassSuppression?: boolean }): void {
    // An upsert that finds no match falls through to addEntry; honor the same
    // suppression bypass on that insert path (relayed task_end upserts, etc.).
    const normalized = normalizeTimelineEntryForStorage(entry);
    if (!normalized) return;

    if (this.tryMergeTaskEndByTaskId(normalized)) {
      return;
    }

    const tolerance = 1000;
    for (let i = this.entries.length - 1; i >= 0; i--) {
      const e = this.entries[i];
      if (e.type === normalized.type && Math.abs(e.ts - normalized.ts) < tolerance) {
        this.entries[i] = {
          ...e,
          raw: normalized.raw,
          ...(normalized.detail ? { detail: normalized.detail } : {}),
          ...(normalized.agentType ? { agentType: normalized.agentType } : {}),
          ...(normalized.projectName ? { projectName: normalized.projectName } : {}),
          ...(normalized.sessionId ? { sessionId: normalized.sessionId } : {}),
          ...(normalized.runId ? { runId: normalized.runId } : {}),
          ...(normalized.taskId ? { taskId: normalized.taskId } : {}),
          ...(normalized.boundarySignal ? { boundarySignal: normalized.boundarySignal } : {}),
          ...(normalized.startedAt != null ? { startedAt: normalized.startedAt } : {}),
          ...(normalized.endedAt != null ? { endedAt: normalized.endedAt } : {}),
          ...(normalized.automated ? { automated: normalized.automated } : {}),
          // summaryKind progresses heuristic/none → llm when the async LLM
          // summary lands. Without this propagation, the dashboard never
          // sees the kind upgrade and (for `summaryKind: 'none'` rows) the
          // detail pane stays suppressed even after the LLM rescues it.
          ...(normalized.summaryKind ? { summaryKind: normalized.summaryKind } : {}),
          ...(normalized.taskScore != null ? { taskScore: normalized.taskScore } : {}),
          ...(normalized.taskOutcome ? { taskOutcome: normalized.taskOutcome } : {}),
          ...(normalized.taskCategory ? { taskCategory: normalized.taskCategory } : {}),
          ...(normalized.taskSummary ? { taskSummary: normalized.taskSummary } : {}),
        };
        for (const cb of this.listeners) cb(this.entries[i], true);
        return;
      }
    }
    this.addEntry(normalized, opts);
  }

  /** Get the most recent entry of a given type */
  getLastEntry(type: string): TimelineEntry | null {
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (this.entries[i].type === type) return this.entries[i];
    }
    return null;
  }

  onEntry(cb: EntryListener): void {
    this.listeners.push(cb);
  }

  removeListener(cb: EntryListener): void {
    const idx = this.listeners.indexOf(cb);
    if (idx !== -1) this.listeners.splice(idx, 1);
  }

  private tryMergeTaskEndByTaskId(entry: TimelineEntry): boolean {
    if (entry.type !== 'task_end' || !entry.taskId) return false;
    let idx = -1;
    for (let i = this.entries.length - 1; i >= 0; i--) {
      const e = this.entries[i];
      if (e.type === 'task_end' && e.taskId === entry.taskId) {
        idx = i;
        break;
      }
    }
    if (idx < 0) return false;

    const existing = this.entries[idx];
    this.entries[idx] = {
      ...existing,
      raw: entry.raw,
      ...(entry.detail ? { detail: entry.detail } : {}),
      ...(entry.agentType ? { agentType: entry.agentType } : {}),
      ...(entry.projectName ? { projectName: entry.projectName } : {}),
      ...(entry.sessionId ? { sessionId: entry.sessionId } : {}),
      ...(entry.runId ? { runId: entry.runId } : {}),
      ...(entry.startedAt != null ? { startedAt: entry.startedAt } : {}),
      ...(entry.endedAt != null ? { endedAt: entry.endedAt } : {}),
      ...(entry.boundarySignal ? { boundarySignal: entry.boundarySignal } : {}),
      ...(entry.summaryKind ? { summaryKind: entry.summaryKind } : {}),
      ...(entry.taskScore != null ? { taskScore: entry.taskScore } : {}),
      ...(entry.taskOutcome ? { taskOutcome: entry.taskOutcome } : {}),
      ...(entry.taskCategory ? { taskCategory: entry.taskCategory } : {}),
      ...(entry.taskSummary ? { taskSummary: entry.taskSummary } : {}),
    };
    for (const cb of this.listeners) cb(this.entries[idx], true);
    return true;
  }
}

File: android/app/src/main/kotlin/dev/agentdeck/state/TimelineStore.kt
package dev.agentdeck.state

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable

@Serializable
data class TimelineEntry(
    val timestamp: Long,
    val type: String,
    val summary: String,
    val detail: String? = null,
    val agentType: String? = null,
    val projectName: String? = null,
    val sessionId: String? = null,
    val runId: String? = null,
    val startedAt: Long? = null,
    val endedAt: Long? = null,
    val status: String? = null,
    val automated: Boolean? = null,
    /** APME task id. Set on task_start/task_end and on every turn entry inside the task scope. */
    val taskId: String? = null,
    /** Only on task_end. todo_complete | clear | session_end | manual | idle_gap. */
    val boundarySignal: String? = null,
    /** "llm" | "heuristic" | "none". Lets clients suppress the detail pane
     *  when the heuristic gave up and the body is just the raw response. */
    val summaryKind: String? = null,
    /**
     * Task-judge rollup, attached on the SECOND `task_end` emit the runner
     * fires once the LLM judge resolves (5–30 s after the boundary). Stores
     * upsert by (type, taskId) — see `upsertEntry`. Nil on the initial emit
     * and for non-task entries / pre-existing on-disk rows. Mirrors the four
     * task* fields on shared/src/timeline.ts::TimelineEntry.
     *   - `taskScore`    : 0..1 composite
     *   - `taskOutcome`  : "success" | "partial" | "fail" | "pending"
     *   - `taskCategory` : task_rollup category
     *   - `taskSummary`  : one-line judge summary
     */
    val taskScore: Double? = null,
    val taskOutcome: String? = null,
    val taskCategory: String? = null,
    val taskSummary: String? = null,
)

data class GroupedEntry(
    val entry: TimelineEntry,
    val count: Int = 1,
    val lastTs: Long = entry.timestamp,
    /** Assistant response body (chat_response) merged into this turn group.
     *  The daemon emits chat_start / chat_response / chat_end as three separate
     *  rows; the UI presents them as one turn row so users see a single
     *  evaluable unit instead of three near-duplicate lines. Mirrors
     *  `apple/AgentDeck/Model/Timeline.swift` GroupedEntry.mergedResponse. */
    val mergedResponse: TimelineEntry? = null,
    /** Terminator metadata (chat_end) merged into this turn group — carries the
     *  "Completed · Ns · topic" suffix. Mirrors Apple mergedCompletion. */
    val mergedCompletion: TimelineEntry? = null,
) {
    /** True when the assistant delivered something for this turn (response body
     *  or completion metadata). UIs use it to stop the chat_start spinner and
     *  swap its icon to the completed glyph — otherwise a merged, completed
     *  turn would spin forever because `chat_start` always maps to the Running
     *  icon. Mirrors Apple GroupedEntry.hasResponse. */
    val hasResponse: Boolean get() = mergedResponse != null || mergedCompletion != null
}

/**
 * Collapse the raw timeline into display groups.
 *
 * Two distinct collapses happen here (mirrors
 * `apple/AgentDeck/Model/Timeline.swift::groupConsecutive`):
 *  1. **Turn merge** — an in-flight `chat_start` absorbs the `chat_response`
 *     and `chat_end` the daemon emits for the same turn, so one user prompt
 *     renders as ONE row instead of three. This is what keeps OpenClaw cron
 *     turns (chat_start + chat_response, each with different text) from
 *     rendering as two rows the way Android used to. A standalone
 *     `chat_response` likewise absorbs a trailing `chat_end`.
 *  2. **Same-type run collapse** — consecutive same-type(+summary) rows fold
 *     into one `×count` group (`canGroup`): tool_request 10s, chat_end/default
 *     60s.
 */
fun groupConsecutive(entries: List<TimelineEntry>): List<GroupedEntry> {
    if (entries.isEmpty()) return emptyList()
    val result = mutableListOf<GroupedEntry>()
    var current = GroupedEntry(entry = entries[0], lastTs = entries[0].timestamp)
    for (i in 1 until entries.size) {
        val entry = entries[i]

        // Task hierarchy markers never group — each is a unique unit of work.
        if (entry.type == "task_start" || entry.type == "task_end" ||
            current.entry.type == "task_start" || current.entry.type == "task_end"
        ) {
            result.add(current)
            current = GroupedEntry(entry = entry, lastTs = entry.timestamp)
            continue
        }

        // Turn merge: chat_start absorbs its chat_response + chat_end. The turn
        // context (`sameTimelineContext` — taskId/runId/sessionId aware, so two
        // turns in one session with distinct runIds never fold together) plus
        // the chat_start anchor (child.startedAt == chat_start.ts) gate the
        // merge — wall-clock does not, because long responses push chat_end far
        // past any window. Synthetic starters ("Prompt sent" / "Connected") are
        // excluded so the trailing completion stays the visible row.
        if (current.entry.type == "chat_start" &&
            current.mergedCompletion == null &&
            isMeaningfulChatStart(current.entry) &&
            sameTimelineContext(current.entry, entry) &&
            sameTurnAnchor(current.entry, entry)
        ) {
            if (entry.type == "chat_response" && current.mergedResponse == null) {
                current = current.copy(mergedResponse = entry)
                continue
            }
            if (entry.type == "chat_end") {
                current = current.copy(mergedCompletion = entry)
                continue
            }
        }

        // A visible chat_response absorbs its trailing chat_end so "Completed ·
        // …" metadata doesn't spawn its own row.
        if (current.entry.type == "chat_response" &&
            current.mergedCompletion == null &&
            entry.type == "chat_end" &&
            sameTimelineContext(current.entry, entry) &&
            sameResponseCompletionAnchor(current.entry, entry)
        ) {
            current = current.copy(mergedCompletion = entry)
            continue
        }

        // Same-type consecutive run collapse (×count) — keep latest entry.
        if (canGroup(current, entry)) {
            current = current.copy(
                entry = entry,
                count = current.count + 1,
                lastTs = entry.timestamp,
            )
        } else {
            result.add(current)
            current = GroupedEntry(entry = entry, lastTs = entry.timestamp)
        }
    }
    result.add(current)
    return result
}

/** True when `child` (chat_response/chat_end) belongs to `start`: the daemon
 *  stamps each child's `startedAt` with the originating chat_start's ts.
 *  Legacy emitters without `startedAt` permit the merge (bounded by iteration
 *  order). Mirrors Apple `sameTurnAnchor`. */
private fun sameTurnAnchor(start: TimelineEntry, child: TimelineEntry): Boolean {
    val anchor = child.startedAt ?: return true
    return anchor == start.timestamp
}

/** Anchor test for chat_response → chat_end absorption. Permits the merge when
 *  either side lacks `startedAt`. Mirrors Apple `sameResponseCompletionAnchor`. */
private fun sameResponseCompletionAnchor(response: TimelineEntry, completion: TimelineEntry): Boolean {
    val r = response.startedAt
    val c = completion.startedAt
    return if (r != null && c != null) r == c else true
}

private fun canGroup(group: GroupedEntry, entry: TimelineEntry): Boolean {
    val prev = group.entry
    if (prev.type != entry.type) return false
    // Task hierarchy markers never group — each task is a unique unit of work.
    if (entry.type == "task_start" || entry.type == "task_end") return false
    val window = when (entry.type) {
        "tool_request" -> 10_000L
        "chat_end" -> 60_000L
        else -> 60_000L
    }
    if (entry.timestamp - group.lastTs > window) return false
    if (!sameTimelineContext(prev, entry)) return false
    // chat_end: group by type only (keep latest summary)
    if (entry.type == "chat_end") return true
    // tool_request: group by type only
    if (entry.type == "tool_request") return true
    // others: same summary
    return prev.summary == entry.summary
}

class TimelineStore private constructor() {

    companion object {
        val instance: TimelineStore by lazy { TimelineStore() }
        private const val MAX_ENTRIES = 500
    }

    private val _entries = MutableStateFlow<List<TimelineEntry>>(emptyList())
    val entries: StateFlow<List<TimelineEntry>> = _entries.asStateFlow()

    fun addEntry(entry: TimelineEntry) {
        val normalized = normalizeTimelineEntryForStorage(entry) ?: return
        // Drop codex:otel-active noise at the **storage** layer so it never
        // reaches consumers that bypass `timelineDisplayGroups` (raw
        // `entries` flow readers, persistence in future, etc.) and so the
        // MAX_ENTRIES buffer doesn't age out useful rows behind it.
        // Mirrors Apple `DaemonTimelineStore` add-path filter.
        val list = _entries.value
        // 5s dedup — skip if same type+summary within window
        for (i in list.indices.reversed()) {
            val e = list[i]
            if (normalized.timestamp - e.timestamp > 5000) break
            if (e.type == normalized.type && e.summary == normalized.summary) return
        }
        _entries.value = (list + normalized).takeLast(MAX_ENTRIES)
    }

    /** Update the most recent entry matching [type] using [transform]. */
    fun updateLastOfType(type: String, transform: (TimelineEntry) -> TimelineEntry) {
        val list = _entries.value.toMutableList()
        val idx = list.indexOfLast { it.type == type }
        if (idx >= 0) {
            list[idx] = transform(list[idx])
            _entries.value = list
        }
    }

    /** Update existing entry with same ts+type (1s tolerance), or add new.
     *
     *  taskId / boundarySignal / summaryKind are progressive: a heuristic
     *  chat_end can later be upserted with summaryKind='llm' + the LLM
     *  summary. Without propagating these, the dashboard keeps showing the
     *  pre-LLM kind and (for 'none' rows) the detail pane stays suppressed
     *  even after the LLM rescues it. */
    fun upsertEntry(entry: TimelineEntry) {
        val normalized = normalizeTimelineEntryForStorage(entry) ?: return
        // Storage-layer guard mirrors addEntry — never let an OTel low-
        // signal row enter, even via the upsert path used for progressive
        // summaryKind enrichment.
        val list = _entries.value.toMutableList()
        // Task-judge follow-up emits land 5–30 s after the initial boundary,
        // so the 1 s timestamp window won't match. For task_end rows, prefer
        // matching by (type, taskId) — that pair is stable across both emits
        // and lets the score-bearing update merge in place. Mirrors Apple
        // `DaemonTimelineStore::upsert`.
        val taskMatchIdx = if (normalized.type == "task_end" && !normalized.taskId.isNullOrEmpty()) {
            list.indexOfLast { it.type == "task_end" && it.taskId == normalized.taskId }
        } else -1
        val idx = if (taskMatchIdx >= 0) taskMatchIdx else {
            list.indexOfLast { it.type == normalized.type && kotlin.math.abs(it.timestamp - normalized.timestamp) < 1000L }
        }
        if (idx >= 0) {
            list[idx] = list[idx].copy(
                summary = normalized.summary,
                detail = normalized.detail ?: list[idx].detail,
                agentType = normalized.agentType ?: list[idx].agentType,
                projectName = normalized.projectName ?: list[idx].projectName,
                sessionId = normalized.sessionId ?: list[idx].sessionId,
                runId = normalized.runId ?: list[idx].runId,
                startedAt = normalized.startedAt ?: list[idx].startedAt,
                endedAt = normalized.endedAt ?: list[idx].endedAt,
                status = normalized.status ?: list[idx].status,
                automated = normalized.automated ?: list[idx].automated,
                taskId = normalized.taskId ?: list[idx].taskId,
                boundarySignal = normalized.boundarySignal ?: list[idx].boundarySignal,
                summaryKind = normalized.summaryKind ?: list[idx].summaryKind,
                taskScore = normalized.taskScore ?: list[idx].taskScore,
                taskOutcome = normalized.taskOutcome ?: list[idx].taskOutcome,
                taskCategory = normalized.taskCategory ?: list[idx].taskCategory,
                taskSummary = normalized.taskSummary ?: list[idx].taskSummary,
            )
            _entries.value = list
        } else {
            addEntry(normalized)
        }
    }

    fun addEntries(newEntries: List<TimelineEntry>) {
        // Same storage-layer guard as addEntry — bulk replay paths
        // (sessions_list snapshots, daemon resync) must not bypass the
        // OTel filter.
        val filtered = newEntries.mapNotNull { normalizeTimelineEntryForStorage(it) }
        _entries.value = (_entries.value + filtered)
            .distinctBy { "${it.timestamp}-${it.type}-${it.summary}" }
            .sortedBy { it.timestamp }
            .takeLast(MAX_ENTRIES)
    }

    /**
     * Replace the entire buffer with a daemon snapshot.
     *
     * The daemon's only `timeline_history` to a dashboard client is the FULL
     * store snapshot sent in `sendInitialState` (Android never issues a
     * session-scoped `query_session_timeline`). Merging it into the existing
     * buffer — the old behaviour — let re-emitted OpenClaw rows stack across
     * reconnects: the daemon re-stamps `ts` on log-tail replay, so the
     * `ts-type-summary` history dedup key shifts and the "same" row survives.
     * The tablet's Wi-Fi reconnect churn made this the dominant duplicate
     * source. Since the snapshot is authoritative, replace rather than merge.
     *
     * Ordering is safe: the daemon sends `connection` → `timeline_history`
     * before any live `timeline_event` on a fresh socket, and TCP preserves
     * order, so no live row is lost by replacing here.
     */
    fun replaceSnapshot(snapshot: List<TimelineEntry>) {
        val filtered = snapshot.mapNotNull { normalizeTimelineEntryForStorage(it) }
        _entries.value = filtered
            .distinctBy { "${it.timestamp}-${it.type}-${it.summary}" }
            .sortedBy { it.timestamp }
            .takeLast(MAX_ENTRIES)
    }

    fun clear() {
        _entries.value = emptyList()
    }
}
